### PR TITLE
✨ Record new tick event fid (first input delay).

### PR DESCRIPTION
**Pull request by ericandrewlewis**
This metric allows tracking the delay that occurs when a user first attempts to interact with a page exposed by the [Event Timing API](https://wicg.github.io/event-timing/).

I haven't introduced a test for this, and would appreciate some feedback on how I should go about this before I do. Should I use sinon to stub/fake out the `PerformanceObserver`, and send fake `firstInput` messages through the faked object?

cc @lannka

### CODE REVIEW

**Comment by lannka** [comment]
@ericandrewlewis yep, an unit test should work. The code looks good!

File: test/unit/test-performance.js [hunk 1]
```diff
@@ -687,3 +687,119 @@ describes.realWin('performance with experiment', {amp: true}, env => {
     });
   });
 });
+
+describes.realWin('performance with firstInput events', {amp: true}, env => {
+  describe('should forward firstInput events as a fid tick event', () => {
+    it('before performance service registered', () => {
+      // Pretend that the EventTiming API exists.
+      env.win.PerformanceEventTiming = true;
```
Comment (ericandrewlewis) [inline test/unit/test-performance.js]: This property gates whether the EventTiming API is enabled or not. I though to stub it so it is automatically cleaned up after the test, but it looks like Sinon does not support this for non-existent object properties.

File: test/unit/test-performance.js [hunk 2]
```diff
@@ -687,3 +687,119 @@ describes.realWin('performance with experiment', {amp: true}, env => {
...
+          .to.be.jsonEqual({
+            label: 'fid',
+            delta: 3,
+          });
+
+      delete env.win.PerformanceEventTiming;
+    });
+
+    // A fake implementation of PerformanceObserver.
+    class PerformanceObserverImpl {
```
Comment (ericandrewlewis) [inline test/unit/test-performance.js]: Is this the best way to fake a class with Sinon?

File: test/unit/test-performance.js [hunk 3]
```diff
@@ -687,3 +687,119 @@ describes.realWin('performance with experiment', {amp: true}, env => {
...
+      });
+
+      // Then registerEventTimingObserver_ will, which is the one we want to
+      // interact with.
+      let performanceObserver;
+      PerformanceObserverStub.onCall(1).callsFake(callback => {
+        performanceObserver = new PerformanceObserverImpl(callback);
+        return performanceObserver;
+      });
+      sinon.replace(env.win, 'PerformanceObserver', PerformanceObserverStub);
```
Comment (ericandrewlewis) [inline test/unit/test-performance.js]: Is this the best way to stub a class constructor?

**Comment by ericandrewlewis** [comment]
Hey @lannka - can you review?

File: src/service/performance-impl.js [hunk 4]
```diff
@@ -213,6 +214,38 @@ export class Performance {
...
+    // That means we need to read existing entries and maintain state
+    // as to whether we have reported a value yet, since in the future it may
+    // be reported twice.
+    // https://bugs.chromium.org/p/chromium/issues/detail?id=725567
+    let recordedFirstInputDelay = false;
+    const processEntry = entry => {
+      if (recordedFirstInputDelay) {
+        return;
+      }
+      this.tick('fid', entry.processingStart - entry.startTime);
```
Comment (ericandrewlewis) [inline src/service/performance-impl.js]: This line is throwing this error in Travis CI:

```
src/service/performance-impl.js:235: ERROR - Property processingStart never defined on entry
      this.tick('fid', entry.processingStart - entry.startTime);
                             ^^^^^^^^^^^^^^^
```

I tried adding an annotation of `PerformanceEntry` for the `entry` parameter like this:

```js
/**
 * @param {PerformanceEntry} entry
 */
```

However, the `processingStart` property is part of the [new `PerformanceEventTiming` interface](https://wicg.github.io/event-timing/#sec-performance-event-timing), and doesn't seem to be defined on the `PerformanceEntry` type. Which makes sense - the Event Timing API is in Editor's Draft status.

I tried specifying an unknown type inline:

```js
this.tick('fid', (/** @type {?} */ (entry)).processingStart - entry.startTime);
```

but the compiler didn't like this either:

```
src/service/performance-impl.js:235: ERROR - Property processingStart never defined on ?
      this.tick('fid', (/** @type {?} */ (entry)).processingStart - entry.startTime);
                                                  ^^^^^^^^^^^^^^^
```

Any suggestion here?

Comment (lannka) [reply src/service/performance-impl.js]: @erwinmombay should this be added to closure compiler extern (I actually don't know where it is) or amp.extern.js?

File: test/unit/test-performance.js (hunk 1 above)
Comment (lannka) [reply test/unit/test-performance.js]: try
```
env.sandbox.stub(env.win, 'PerformanceObserver', stub)
```

we create `sinon.sandbox` object for each window context, which will be reset for each test case

**Comment by calebcordry** [comment]
Should we add this new tick event to `TICKEVENTS.md`?

File: src/service/performance-impl.js (hunk 4 above)
Comment (calebcordry) [inline src/service/performance-impl.js]: Also looks like it is intended for you call `tickDelta` if you are passing the second argument (I know it just calls tick 🤷‍♂️)
https://github.com/ampproject/amphtml/blob/b84350059273cec5d8e683b3bfae073cf356dcd1/src/service/performance-impl.js#L309-L311

File: src/service/performance-impl.js [hunk 5]
```diff
@@ -213,6 +214,38 @@ export class Performance {
...
+      recordedFirstInputDelay = true;
+    };
+    const observer = new this.win.PerformanceObserver(list => {
```
Comment (calebcordry) [inline src/service/performance-impl.js]: Should we try to share the instance of `PerformanceObserver` across this and the `paintTimingObserver`?

File: test/unit/test-performance.js (hunk 3 above)
Comment (calebcordry) [reply test/unit/test-performance.js]: Looks like @lannka answered this in the comment above.

Comment (ericandrewlewis) [reply test/unit/test-performance.js]: Done [this](https://github.com/ampproject/amphtml/pull/20721#discussion_r256158017)

File: src/service/performance-impl.js (hunk 5 above)
Comment (ericandrewlewis) [reply src/service/performance-impl.js]: Yeah let's do it - we'll have more perf metrics to record in the future. Done.

File: src/service/performance-impl.js (hunk 4 above)
Comment (ericandrewlewis) [reply src/service/performance-impl.js]: I created a separate extern for these definitions for now. 

[Maybe](https://github.com/google/closure-compiler/issues/3233) this should be added to closure-compiler itself - I'll follow-up there if so. Defining our own extern in this project allows us to move forward with this feature for the moment.

**Comment by ericandrewlewis** [comment]
> Should we add this new tick event to `TICKEVENTS.md`?

Sure - I added documentation for that.

@lannka / @calebcordry this is now ready for re-review.

File: build-system/tasks/compile.js [hunk 6]
```diff
@@ -118,6 +118,7 @@ function compile(entryModuleFilenames, outputDir, outputFilename, options) {
   const baseExterns = [
     'build-system/amp.extern.js',
     'build-system/dompurify.extern.js',
+    'build-system/event-timing.extern.js',
```
Comment (calebcordry) [inline build-system/tasks/compile.js]: @erwinmombay can you please review externs?

File: test/unit/test-performance.js [hunk 7]
```diff
@@ -687,3 +687,207 @@ describes.realWin('performance with experiment', {amp: true}, env => {
...
+      };
+      // Fake a triggering of the firstInput event.
+      performanceObserver.triggerCallback(list);
+      expect(perf.events_.length).to.equal(1);
+      expect(perf.events_[0])
+          .to.be.jsonEqual({
+            label: 'fid',
+            delta: 3,
+          });
+      delete env.win.PerformanceEventTiming;
```
Comment (lannka) [inline test/unit/test-performance.js]: consider put the delete into `afterEach`

Comment (ericandrewlewis) [reply test/unit/test-performance.js]: Hmm - one of the tests fakes `env.win.PerformancePaintTiming` and the three others fake `env.win.PerformanceEventTiming`. We'll introduce other faked properties in this manner (like `window.PerformanceLayoutJank` so I think leaving these in the test cases is okay?

Comment (lannka) [reply test/unit/test-performance.js]: Ah, didn't realize they're different.

**Comment by rsimha** [comment]
@lannka @ericandrewlewis One of the tests added by this PR is routinely failing on Travis. For example, https://travis-ci.org/ampproject/amphtml/jobs/498416589#L2353-L2371

```
      IT => created before performance service registered
        ✗ entryTypes contained only unsupported types
        observe@[native code]
        registerPerformanceObserver_@/home/travis/build/ampproject/amphtml/src/service/performance-impl.js:233:21 <- /tmp/5502fab4fcfd3899ef4e381a680b0b8b.browserify.js:88542:21
        Performance@/home/travis/build/ampproject/amphtml/src/service/performance-impl.js:115:4 <- /tmp/5502fab4fcfd3899ef4e381a680b0b8b.browserify.js:88419:38
        getServiceInternal@/home/travis/build/ampproject/amphtml/src/service.js:401:23 <- /tmp/5502fab4fcfd3899ef4e381a680b0b8b.browserify.js:77007:23
        /home/travis/build/ampproject/amphtml/test/unit/test-performance.js:745:43 <- /tmp/5502fab4fcfd3899ef4e381a680b0b8b.browserify.js:146239:51
```

Let me know if this is a quick fix. If not, I can send a PR to skip this test while you investigate.

**Pull request merged**